apps/rush-lib/src/cli/actions/TabCompleteAction.ts

```diff
@@ -0,0 +1,218 @@
+              }
+            } else {
+              if (projectCommands.indexOf(lastCommand) !== -1) {
+                for (let i: number = 0; i < this.rushConfiguration.projects.length; i++) {
+                  yield this.rushConfiguration.projects[i].packageName;
+                }
+
+                return;
+              }
+            }
+            // TODO: Add support for version policy, variant
+          } else if (actionName === 'change') {
+            const bumpTypes: string[] = ['major', 'minor', 'patch', 'none'];
+            if (completePartialWord) {
+              if (secondLastCommand === '--bump-type') {
+                for (let i: number = 0; i < bumpTypes.length; i++) {
+                  if (bumpTypes[i].indexOf(lastCommand) === 0) {
+                    yield bumpTypes[i];
+                  }
+                }
+
+                return;
+              }
+            } else {
+              if (lastCommand === '--bump-type') {
+                for (let i: number = 0; i < bumpTypes.length; i++) {
+                  yield bumpTypes[i];
+                }
+
+                return;
+              }
+            }
+          } else if (actionName === 'publish') {
+            const accessLevels: string[] = ['public', 'restricted'];
+            if (completePartialWord) {
+              if (secondLastCommand === '--set-access-level') {
+                for (let i: number = 0; i < accessLevels.length; i++) {
+                  if (accessLevels[i].indexOf(lastCommand) === 0) {
+                    yield accessLevels[i];
+                  }
+                }
+
+                return;
+              }
+            } else {
+              if (lastCommand === '--set-access-level') {
+                for (let i: number = 0; i < accessLevels.length; i++) {
+                  yield accessLevels[i];
+                }
+
+                return;
+              }
+            }
+          }
+
+          if (completePartialWord) {
+            for (let i: number = 0; i < TabCompleteAction._actions[actionName].length; i++) {
+              if (TabCompleteAction._actions[actionName][i].name.indexOf(lastCommand) === 0) {
+                yield TabCompleteAction._actions[actionName][i].name;
+              }
+            }
+          } else {
+            for (let i: number = 0; i < TabCompleteAction._actions[actionName].length; i++) {
+              if (
+                lastCommand === TabCompleteAction._actions[actionName][i].name &&
+                TabCompleteAction._actions[actionName][i].kind !== CommandLineParameterKind.Flag
+              ) {
+                // The parameter is expecting a value, so don't suggest parameter names again
+                return;
+              }
+            }
+
+            for (let i: number = 0; i < TabCompleteAction._actions[actionName].length; i++) {
+              yield TabCompleteAction._actions[actionName][i].name;
+            }
+          }
+        }
+      }
+    }
```

Comment:
This really should be autogenerated.

Reply:
I looked at that but couldn't find a way to get the set of possible values for choice parameters from the parser object. I can optimize this part a bit, though.

Reply:
We can expose that better from the ts-command-line API if needed.

Reply:
In principle the entire CLI schema should be readable via the API. We should not rely on the `argparse` object for this, because we're going to get rid of that soon. Instead the TypeScript classes should simply store their own inputs.